File: projects/08/clojure/src/Temp.clj
;; Temp.clj is a module for our 'Temp' type-abstraction.
;;
;; Begun by: Prof. Adams, for CS 214 at Calvin College.
;; Completed by: Sinai Park (sp46)
;; Date: April 15 2020
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

;; Create a record/struct type named Temp
(defrecord Temp [degree, scale])


;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; make-Temp constructs a Temp object from 
;; degree, a string and scale, a char 
;; Receive: degree and scale
;; Return: the Temp (degree and scale). 
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(defn make-Temp [degree scale]
  (->Temp degree scale)
)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; getDegrees extracts the degrees of a temp object.
;; Receive: aTemp, a Temp
;; Return: the degrees double in aTemp.
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(defn getDegree [aTemp]
  (:degree aTemp)
)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; getScale extracts the scale of a temp object.
;; Receive: aTemp, a Temp
;; Return: the scale char in aTemp.
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(defn getScale [aTemp]
  (:scale aTemp)
)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; toFarenheit extracts the degrees of a temp object.
;; Receive: aTemp, a Temp
;; Return: the degrees in Fahrenheit
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(defn toFahrenheit [aTemp]
  (if (or (=(:scale aTemp) "C") (=(:scale aTemp) "c"))
    (do
      (let [(:degree aTemp) (+ ( * (:degree aTemp) (/ 9.0 5.0)) 32.0) ]
      (let [scale_char ("F")]
      (->Temp degree_val scale_char)))))
  (if (or (=(:scale aTemp) "K") (=(:scale aTemp) "k"))
    (do
      (let [degree_val (- ( * (:degree aTemp) (/ 9.0 5.0)) 459.67) ]
      (let [scale_char ("F")]
      (->Temp degree_val scale_char)))))
)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; toCelcius extracts the scale of a temp object.
;; Receive: aTemp, a Temp. 
;; Return: the degrees in Celsius
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(defn toCelsius [aTemp]
  (:scales aTemp)
)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; toKelvin extracts the scale of a temp object.
;; Receive: aTemp, a Temp
;; Return: the degrees in Kelvin
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(defn toKelvin [aTemp]
  (:scales aTemp)
)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;; readTemp() creates a temp from user input. 
;;; Receive: aTemp, a Temp 
;;; Return: none 
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
 
(defn readTemp[^Temp aTemp]
  (print "Please enter the degree name: ") (flush)
  (let
    [degree (read-line)]
      (print "Please enter the scale: ") (flush)
      (let
        [ scale (read-line)]
          (->Temp degree scale)
          
      )
  )
)
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; printTemp displays a temp object. 
;; Receive: aTemp, a Temp, 
;; Output: the string representation of aTemp. 
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(defn printTemp[aTemp]
  (print (getDegree aTemp) (getScale aTemp))
)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;; setFirst() changes the first name of a Name object.
;;; Receive: aName, a Name.
;;; Return: the changed firstName string in aName.
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(defn raise[aTemp degreeval]
    (->Temp (+ (getDegree aTemp) degreeval) (:scale aTemp) )
  
)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;; setMiddle() changes the middle name of a name object.
;;; Receive: aName, a Name. 
;;; Return: the middleName string in aName.
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
 
(defn lower[aTemp sub_degree]
  (->Temp (- (:degree aTemp) sub_degree) (:scale aTemp) )
)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;; setLast() changes the last name of a name object.
;;; Receive: aName, a Name. 
;;; Return: the lastName string in aName.
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
 
(defn equals[aTemp comp_temp]
  (:degree aTemp)
)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;; less_than() returns a full name in L-F-M order. 
;;; Receive: aName, a Name. 
;;; Return: lastName, firstName, middleName,
;;;           separated by spaces. 
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
 
(defn less_than[aTemp comp_temp]
  (:degree aTemp)
)

File: projects/08/java/Temp.java
/* Temp.java implements class Temperature.
 * Begun by: Prof. Adams, for CS 214 at Calvin College.
 * Student: Sinai Park(sp46)
 * Date: April 10 2020
 ************************************************************/


import java.io.*;
import java.util.Scanner;

class Temp
   {

      private double degrees;
      private char   scales;              

     /***************************************************
     * Temp constructs a name from three strings.      *
     * Receive: first, middle and last, three strings. *
     * Return: the triplet (first middle last).        *
     ***************************************************/
      public Temp(double degrees, char scale)
      {
        degrees = degrees;
        scales = scale;
      }
      //getDegrees() returns the degree value of the Temp object          
      public double getDegrees()
      {
         return degrees;
      }
      //getScale() returns the scale value of the Temp object
      public char getScale()
      {
         return scales;
      }

      //function toFarenheit() to return the temp to f
      public void toFarenheit() 
      {
      if (scales == 'c' || scales == 'C'){
	 degrees = (getDegrees() * 9.0/5.0) + 32.0;
	 scales = 'F';}
      else if (scales == 'K' || scales == 'k'){
	 degrees = (getDegrees() - 273.15) * (9.0/5.0) + 32.0;
	 scales = 'F';}
	}
     
      //function toCelcius() to return the temp to c
      public void toCelcius() 
      {
      if (scales == 'K' || scales == 'k')
         {
	 degrees = getDegrees() - 273.15;
         scales = 'C';
	 }
      else if (scales == 'f' || scales == 'F')
         {
	 degrees = (getDegrees() - 32.0) * (5.0/9.0);
         scales = 'C';
	 }
	}

      //function toKelvin() to return the temp to k
      public void toKelvin() 
      {
      if (scales == 'c' || scales == 'C'){
	 degrees = getDegrees() + 273.15;
	 scales = 'K';}
      else if (scales == 'F' || scales == 'f'){
	 degrees = ( getDegrees() - 32 ) * (5.0/9.0) + 273.15;
	 scales = 'K';}
	}

      //function readTemp() reads in the given value of temp from the user
      public void readTemp(Scanner keyboard)
      {
	 System.out.println("Please enter the temperature value with scale: ");
	 String temp = keyboard.nextLine();
	 String[] theArray = temp.split(" ");

	String degree_val = theArray[0];
	double degree_valid = Double.parseDouble(degree_val);

	String scale_char = theArray[1];
	char scale_valid = scale_char.charAt(0);

	if ( isValid(degree_valid, scale_valid) )
		{
			degrees = degree_valid;
			scales = scale_valid;
		}
		else {
			System.out.println("Invalid temperature");
		}
      }
	//function printTemp() reads in the given value of temp from the user
      public void printTemp()
      {
	 double degree_string = getDegrees();
	 System.out.print(Double.toString(degree_string) + " " + getScale() );
	 

      }
      //raise() doubles the amount of given degrees
      public void raise(double degree) 
      {
         degrees = degrees + degree;
      }

	//lower() doubles the amount of given degrees
	public void lower(double degree)
	{
		if ( isValid(degrees - degree, scales) )
			degrees = degrees - degree;
		else
			System.out.println("New temperature is invalid!");
	}

	//function isValid takes in a degree and scale to see if it is a valid temperature
	public boolean isValid(double Degrees, char Scale)
	{
		if ( Scale == 'K' || Scale == 'k' )
		{
			if ( Degrees < 0.0 )
				return false;
			else
				return true;
		}
		else if ( Scale == 'C' || Scale == 'c' )
		{
			if ( Degrees < -237.0 )
				return false;
			else
				return true;
		}
		else if ( Scale == 'F' || Scale == 'f' )
		{
			if ( Degrees < -459.0 )
				return false;
			else
				return true;
		}
		else
			return false;
	}
	
	//function equals() checks if the two temperatures are equal regardless of scale
	public boolean equals(Temp comp) 
	{
	char my_scale = getScale();

	//set both degrees to farenheight
	toFarenheit();
	comp.toFarenheit();

	//compare the degree values
	if (comp.getDegrees() == getDegrees() )
	{
	 if (my_scale == 'C' || my_scale == 'c') { 
		toCelcius();}
	else if (my_scale == 'K' || my_scale == 'k') { 
		toKelvin();}
	return true;
	} 
	else 
		return false;
	}

	//function less_than() checks if the left temperature is less than the right temp regardless of scale
	public boolean less_than(Temp comp) 
	{
	char my_scale = getScale();
	//set both degrees to farenheight
	toFarenheit();
	comp.toFarenheit();

	if (getDegrees() < comp.getDegrees() )
		{
		if (my_scale == 'C' || my_scale == 'c') { 
			toCelcius();}
		else if (my_scale == 'K' || my_scale == 'k') { 
			toKelvin();}
		return true; 
		}
		else
			return false;
	
	}
}
